src/sources/aws_s3/sqs.rs

```diff
@@ -0,0 +1,517 @@
+                        }
+                    }
+
+                    futures::future::ready(Some(Ok(event)))
+                });
+
+                out.send_all(Compat::new(Box::pin(stream)))
+                    .compat()
+                    .await
+                    .map_err(|error| {
+                        error!(message = "Error sending S3 Logs", %error);
+                    })
+                    .ok();
+
+                read_error
+                    .map(|error| {
+                        Err(ProcessingError::ReadObject {
+                            error,
+                            bucket: s3_event.s3.bucket.name.clone(),
+                            key: s3_event.s3.object.key.clone(),
+                        })
+                    })
+                    .unwrap_or(Ok(()))
+            }
+            None => Ok(()),
+        }
+    }
+
+    async fn receive_messages(&self) -> Result<Vec<Message>, RusotoError<ReceiveMessageError>> {
+        self.sqs_client
+            .receive_message(ReceiveMessageRequest {
+                queue_url: self.queue_url.clone(),
+                max_number_of_messages: Some(10),
+                visibility_timeout: Some(self.visibility_timeout_secs),
+                ..Default::default()
+            })
+            .map_ok(|res| res.messages.unwrap_or_default())
+            .await
+    }
+
+    async fn delete_message(
+        &self,
+        receipt_handle: String,
+    ) -> Result<(), RusotoError<DeleteMessageError>> {
+        self.sqs_client
+            .delete_message(DeleteMessageRequest {
+                queue_url: self.queue_url.clone(),
+                receipt_handle,
+            })
+            .await
+    }
+}
+
+// https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "PascalCase")]
+struct S3Event {
+    records: Vec<S3EventRecord>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct S3EventRecord {
+    event_version: String, // TODO compare >=
+    event_source: String,
+    aws_region: String,
+    event_name: S3EventName,
+
+    s3: S3Message,
+}
+
+#[derive(Clone, Debug)]
+struct S3EventName {
+    kind: String,
+    name: String,
+}
+
+// https://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html#supported-notification-event-types
+//
+// we could enums here, but that seems overly brittle as deserialization would  break if they add
+// new event types or names
+impl<'de> Deserialize<'de> for S3EventName {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        let s = String::deserialize(deserializer)?;
+
+        let mut parts = s.splitn(2, ':');
+
+        let kind = parts
+            .next()
+            .ok_or_else(|| D::Error::custom("Missing event type"))?
```

Comment:
Event "type" or "kind"?

Reply:
Good call, I wanted to use `type` but given it is a reserved keyword, I shifted to `kind`. I'll update this.